mmv1/templates/cai2hcl/resource_converter.go.erb

```diff
@@ -0,0 +1,138 @@
+
+const <%=resource_name-%>AssetType string = "<%=product_backend_name.downcase-%>.googleapis.com/<%=object.name-%>"
+
+type <%=converter_name-%> struct {
+	name   string
+	schema map[string]*tfschema.Schema
+}
+
+func New<%=converter_name-%>(name string) cai2hcl.Converter {
+	schema := tpg.Provider().ResourcesMap[name].Schema
+
+	return &<%=converter_name-%>{
+		name:   name,
+		schema: schema,
+	}
+}
+
+func (c *<%=converter_name-%>) Convert(assets []*caiasset.Asset) ([]*cai2hcl.HCLResourceBlock, error) {
+	var blocks []*cai2hcl.HCLResourceBlock
+	config, error := cai2hclConfig.NewConfig()
+	if error != nil {
+		return nil, error
+	}
+
+	for _, asset := range assets {
+		if asset == nil {
+			continue
+		}
+		if asset.Resource != nil && asset.Resource.Data != nil {
+			block, err := c.convertResourceData(asset, config)
+			if err != nil {
+				return nil, err
+			}
+			blocks = append(blocks, block)
+		}
+	}
+	return blocks, nil
+}
+
+func (c *<%=converter_name-%>) convertResourceData(asset *caiasset.Asset, config *transport_tpg.Config) (*cai2hcl.HCLResourceBlock, error) {
+	if asset == nil || asset.Resource == nil || asset.Resource.Data == nil {
+		return nil, fmt.Errorf("asset resource data is nil")
+	}
+
+	var resource *<%=api_type-%>
+
+	if err := cai2hclHelper.DecodeJSON(asset.Resource.Data, &resource); err != nil {
+		return nil, err
+	}
+	
+    hcl, _ := resource<%=resource_name-%>Read(resource, config)
+
+	ctyVal, err := cai2hclHelper.MapToCtyValWithSchema(hcl, c.schema)
+	if err != nil {
+		return nil, err
+	}
+	return &cai2hcl.HCLResourceBlock{
+		Labels: []string{c.name, resource.Name},
+		Value:  ctyVal,
+	}, nil
+}
+
+func resource<%=resource_name-%>Read(res *<%=api_type-%>, config map[string]*tfschema.Schema) (map[string]interface{}, error) {
+    result := make(map[string]interface{})
+
+<%  object.gettable_properties.reject{|p| p.ignore_read }.each do |prop|-%>
+<%    if prop.flatten_object-%>
+    if flattenedProp := flatten<%="Nested" if object.nested_query-%><%=resource_name-%><%=titlelize_property(prop)-%>(res.<%= prop.api_name.camelize(:upper) -%>); flattenedProp != nil {
+        if gerr, ok := flattenedProp.(*googleapi.Error); ok {
+			return nil, fmt.Errorf("Error reading <%=object.name-%>: %s", gerr)
+		}
+        casted := flattenedProp.([]interface{})[0]
+        if casted != nil {
+            for k, v := range casted.(map[string]interface{}) {
+                result[k] = v
+            }
+        }
+    }
+<%    else-%>
+    result["<%=prop.name.underscore-%>"] = flatten<%="Nested" if object.nested_query-%><%=resource_name-%><%=titlelize_property(prop)-%>(res.<%= prop.api_name.camelize(:upper) -%>, config)
+<%    end-%>
+<%  end-%>
+
+<%  if object.has_self_link-%>
+    result["self_link"] = tpgresource.ConvertSelfLinkToV1(res.SelfLink)
+<%  end-%>
+
+    return result, nil
+}
+
+<%- nested_prefix = object.nested_query ? "Nested" : ""-%>
+<% object.gettable_properties.reject(&:ignore_read).each do |prop|-%>
```

Comment:
`ignore_read` properties are ones that can be set but not read - we probably want to include them in this conversion code if possible?

Reply:
(but it's okay to leave them out as a first pass)

Reply:
In future yeah, we'll need to address it, its not perfectly clear for me how to do it yet as flatten_property_method itself [ignores](https://github.com/GoogleCloudPlatform/magic-modules/blob/642e4540b7055efa319de796b95bb03206131729/mmv1/templates/terraform/flatten_property_method.erb#L75C47-L75C58) such fields: